### PR TITLE
WIP: Forecast Hub Updater

### PR DESCRIPTION
**Pull request by BrettBoval**
Caching the Forecast Hub Ensemble Forecast into covid-data-public

### CODE REVIEW

**Comment by BrettBoval** [comment]
@ghop02 the details of column name mapping and shape still need to be standardized, but can you take a look at the skeleton and let me know what you think. I based it on the update_nytime_data.py script.

**Comment by BrettBoval** [comment]
@ghop02 PTAL

File: scripts/update_forecast_hub.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        ]
+        mask = np.logical_and.reduce(masks)
+
+        # The raw data is in long form and we need to pivot this to create a column for
+        # WEEKLY_NEW_CASES and WEEKLY_NEW_DEATHS. "target_type" has either death or cases. "value"
+        # has the predicted value. The rest of the columns create a unique index. For right now only
+        # one model and one forecast_date are being served, but we need to maintain the option of
+        # multiple values.
+        COLUMNS = [
+            "model_abbr",
```
Comment (ghop02) [inline scripts/update_forecast_hub.py:140]: Would prefer `Fields.MODEL_ABBR`, etc here!

File: scripts/update_forecast_hub.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    GOOGLE = "Google_Harvard-CPF"
+
+
+class Fields(GetByValueMixin, FieldNameAndCommonField, enum.Enum):
+    MODEL_ABBR = "model_abbr", CommonFields.MODEL_ABBR
+    REGION = "unit", CommonFields.FIPS
+    FORECAST_DATE = "forecast_date", CommonFields.FORECAST_DATE
+    TARGET_DATE = "target_date", CommonFields.DATE
+    QUANTILE = "quantile", CommonFields.QUANTILE
+    NEW_CASES = "case", CommonFields.WEEKLY_NEW_CASES
```
Comment (ghop02) [inline scripts/update_forecast_hub.py:44]: Can you walk me through your thought process of `WEEKLY_NEW_CASES` vs `NEW_CASES`? My initial reaction was it feels a bit weird to bind to a specific time frame

File: scripts/update_forecast_hub.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        )
+
+    @property
+    def raw_path(self):
+        return self.raw_data_root / self.RAW_CSV_FILENAME
+
+    def write_version_file(self, forecast_date) -> None:
+        stamp = datetime.datetime.utcnow().isoformat()
+        version_path = self.raw_data_root / "version.txt"
+        with version_path.open("w+") as vf:
```
Comment (ghop02) [inline scripts/update_forecast_hub.py:81]: Will this append to the existing file as opposed to replacing it?

File: covidactnow/datapublic/common_fields.py [hunk 4]
```diff
@@ -44,6 +44,19 @@ class CommonFields(GetByValueMixin, ValueAsStrMixin, str, Enum):
     CASES = "cases"
     DEATHS = "deaths"
     RECOVERED = "recovered"
+
+    # Incidence Values
```
Comment (ghop02) [inline covidactnow/datapublic/common_fields.py:48]: Incident values?

File: scripts/update_forecast_hub.py (hunk 3 above)
Comment (BrettBoval) [reply scripts/update_forecast_hub.py:81]: 'a' appends. In this case 'w' and 'w+' both overwrite the existing file. I had it at 'w+' for consistency with the update_nytimes_data.py script but switched it to 'w' to stop the confusion.

File: covidactnow/datapublic/common_fields.py (hunk 4 above)
Comment (BrettBoval) [reply covidactnow/datapublic/common_fields.py:48]: done.

File: requirements.txt [hunk 5]
```diff
@@ -1,3 +1,5 @@
+zoltpy
+numpy
```
[the remark this answers is not in the document]
Comment (BrettBoval) [reply requirements.txt:2]: done

File: scripts/update_forecast_hub.py (hunk 1 above)
Comment (BrettBoval) [reply scripts/update_forecast_hub.py:140]: done

File: scripts/update_forecast_hub.py (hunk 2 above)
Comment (BrettBoval) [reply scripts/update_forecast_hub.py:44]: The Forecast Hub explicitly only accepts, stores, and serves weekly incidence values for cases and deaths ("how many deaths in the next week"). I wanted to make it clear that these are weekly rates as opposed to daily values that are intermittently provided. I think CommonFields.NEW_CASES should be for daily new cases.

**Comment by ghop02** [comment]
Can you revert the changes to the data files? We try to only update the files in data (with the exception of new files) through the github action

File: scripts/update_forecast_hub.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        wide_df = data.set_index(
+            [
+                Fields.REGION.common_field,
+                Fields.TARGET_DATE.common_field,
+                Fields.MODEL_ABBR.common_field,
+                Fields.FORECAST_DATE.common_field,
+            ]
+        ).pivot(columns=Fields.QUANTILE.common_field)
+        wide_df.columns = wide_df.columns = [
+            x[0] + "_" + str(x[1]) for x in wide_df.columns.to_flat_index()
```
Comment (ghop02) [inline scripts/update_forecast_hub.py:181]: Do we need all the quantiles?  I was envisioning having a couple column names like "WEEKLY_NEW_CASES_90", "WEEKLY_NEW_CASES_10" as the 90th/10th quantile.  It's definitely a bit more verbose but just in the name of having explicit column names.

Comment (BrettBoval) [reply scripts/update_forecast_hub.py]: I added a TODO. The requirements on which quantiles will be needed is still shifting. Once that settles I'm in favor of explictly filtering and renaming only a subset of these.

File: scripts/update_forecast_hub.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        # Using common_field because this is done after helpers.rename_fields
+        wide_df = data.set_index(
+            [
+                Fields.REGION.common_field,
+                Fields.TARGET_DATE.common_field,
+                Fields.MODEL_ABBR.common_field,
+                Fields.FORECAST_DATE.common_field,
+            ]
+        ).pivot(columns=Fields.QUANTILE.common_field)
+        wide_df.columns = wide_df.columns = [
```
[the remark this answers is not in the document]
Comment (BrettBoval) [reply scripts/update_forecast_hub.py]: fixed. that's a new error for me.

**Comment by BrettBoval** [comment]
> Can you revert the changes to the data files? We try to only update the files in data (with the exception of new files) through the github action

Done. LMK if there's a better way to test the YAML github action files or if I'm just supposed to go in and comment out the final commit section.

File: requirements.txt [hunk 8]
```diff
@@ -15,3 +16,5 @@ s3fs==0.4.2
 boto3==1.13.26
 click==7.1.2
 xlrd==1.2.0
+git+https://github.com/reichlab/zoltpy/
```
Comment (ghop02) [inline requirements.txt:19]: Can you pin this to either a commit or explicitly master?

**Review by ghop02 — APPROVED** [review]

**Pull request merged**